asciidtype/asciidtype/src/casts.c

```diff
@@ -0,0 +1,180 @@
+
+    return 0;
+}
+
+static int
+ascii_to_ascii_strided(PyArrayMethod_Context *context, char *const data[],
+                       npy_intp const dimensions[], npy_intp const strides[],
+                       NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_Descr **descrs = context->descriptors;
+    long in_size = ((ASCIIDTypeObject *)descrs[0])->size;
+    long out_size = ((ASCIIDTypeObject *)descrs[1])->size;
+    long copy_size;
+
+    if (out_size > in_size) {
+        copy_size = in_size;
+    }
+    else {
+        copy_size = out_size;
+    }
+
+    npy_intp N = dimensions[0];
+    char *in = data[0];
+    char *out = data[1];
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+
+    while (N--) {
+        for (int i = 0; i < copy_size; i++) {
+            *(out + i) = *(in + i);
+        }
+        for (int i = copy_size; i < out_size; i++) {
+            *(out + i) = '\0';
+        }
+        in += in_stride;
+        out += out_stride;
+    }
+
+    return 0;
+}
+
+static int
+ascii_to_ascii_unaligned(PyArrayMethod_Context *context, char *const data[],
+                         npy_intp const dimensions[], npy_intp const strides[],
+                         NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_Descr **descrs = context->descriptors;
+    long in_size = ((ASCIIDTypeObject *)descrs[0])->size;
+    long out_size = ((ASCIIDTypeObject *)descrs[1])->size;
+    long copy_size;
+
+    if (out_size > in_size) {
+        copy_size = in_size;
+    }
+    else {
+        copy_size = out_size;
+    }
+
+    npy_intp N = dimensions[0];
+    char *in = data[0];
+    char *out = data[1];
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+
+    while (N--) {
+        memcpy(out, in, out_size * sizeof(char));  // NOLINT
+        for (int i = copy_size; i < out_size; i++) {
+            *(out + i) = '\0';
+        }
+        in += in_stride;
+        out += out_stride;
+    }
+
+    return 0;
+}
+
+static int
+ascii_to_ascii_get_loop(PyArrayMethod_Context *context, int aligned,
+                        int NPY_UNUSED(move_references),
+                        const npy_intp *strides,
+                        PyArrayMethod_StridedLoop **out_loop,
+                        NpyAuxData **NPY_UNUSED(out_transferdata),
+                        NPY_ARRAYMETHOD_FLAGS *flags)
+{
+    PyArray_Descr **descrs = context->descriptors;
+
+    int contig = (strides[0] == ((ASCIIDTypeObject *)descrs[0])->size *
+                                        sizeof(char) &&
+                  strides[1] == ((ASCIIDTypeObject *)descrs[1])->size *
+                                        sizeof(char));
+
+    if (aligned && contig) {
+        *out_loop = (PyArrayMethod_StridedLoop *)&ascii_to_ascii_contiguous;
+    }
+    else if (aligned) {
+        *out_loop = (PyArrayMethod_StridedLoop *)&ascii_to_ascii_strided;
+    }
+    else {
+        *out_loop = (PyArrayMethod_StridedLoop *)&ascii_to_ascii_unaligned;
```

Comment:
The aligment of `char` will be 1, so while I think I currently sometimes pass "unaligned" to not bother figuring out alignment, you could just use the same loop for both.  I suppose you can also argue that char isn't standardized to 1 byte :).

Reply:
Its hard, but there are four paths right now how you can reach unaligned casts:
* I think I had some scalar paths where I didn't bother checking alignment and just calling the unaligned version).  (I have to double check this, but I think it happens for casts between dtypes when assigning a single value.)
* You can manage it by manually ensuring you have unaligned data ;).  But that is tricky, e.g. with `np.zeros(1001)[1:].view("U10")`.
* Structured dtypes by default use packed structs rather than aligned ones.  So if you do `dtype=[("a", "?"), ("b", "U5")]` than the `"b"` field is unaligned.
* Structured dtypes are a mess and they (currently) do not even try to pass on correct alignment to the field dtypes when necessary (i.e. in casts).  That is something that can be done, but I really would like to pass through the *actual* alignment, rather than just "its all aligned".  (https://github.com/numpy/numpy/issues/17359)

The last point is the main one really, the second one is rather theoretical...  Which is why my suggestion to allow flagging "do not allow unaligned".

Reply:
Thanks! Having such a flag, with the default to not allow aunaligned, would certainly reduce the conceptual overhead quite a bit.